### PR TITLE
Web searcher

### PR DESCRIPTION
**Pull request by giann**

### CODE REVIEW

**Comment by daurnimator** [comment]
Typo in commit message: "REAMDE".

File: src/defs.js [hunk 1]
```diff
@@ -297,6 +297,12 @@ if (WEB) {
     const LUA_CPATH_DEFAULT =
         LUA_CDIR + "?.js;" + LUA_CDIR + "loadall.js;./?.js";
     module.exports.LUA_CPATH_DEFAULT = LUA_CPATH_DEFAULT;
+
+    const LUA_WEB_PATH_DEFAULT = "?.lua;?/init.lua";
```
Comment (daurnimator) [inline src/defs.js]: We shouldn't need new variables for this. `if(WEB)` should mean that `LUA_CPATH_DEFAULT` is already this.

File: src/lauxlib.js [hunk 2]
```diff
@@ -644,19 +644,72 @@ const luaL_unref = function(L, t, ref) {
...
+    constructor() {
+        this.n = NaN;  /* number of pre-read characters */
+        this.f = null;  /* file being read */
+        this.buff = new Buffer(1024);  /* area for reading file */
+        this.pos = 0;  /* current position in file */
+        this.err = void 0;
     }
+}
+
+if (!WEB) {
```
Comment (daurnimator) [inline src/lauxlib.js]: to be consistent with elsewhere in the codebase, lets put the `if(WEB)` case first

File: src/lauxlib.js [hunk 3]
```diff
@@ -783,6 +787,70 @@ if (!WEB) {
...
+            lua.lua_pushliteral(L, `@${jsfilename}`);
+
+            let xhr = new XMLHttpRequest();
+            xhr.open("GET", jsfilename, false);
+            //xhr.responseType = "arraybuffer";
+            xhr.send();
+
+            if (xhr.status === 200) {
+                /* TODO: Synchronous xhr alway return a js string */
+                lf.f = lua.to_luastring(xhr.response);
```
Comment (daurnimator) [inline src/lauxlib.js]: Doesn't this need to set `lf.buff`?

File: src/lauxlib.js [hunk 4]
```diff
@@ -783,6 +787,70 @@ if (!WEB) {
...
+        let status = lua.lua_load(L, getF, lf, lua.lua_tostring(L, -1), mode);
+        let readstatus = lf.err;
+        if (readstatus) {
+            lua.lua_settop(L, fnameindex);  /* ignore results from 'lua_load' */
+            return errfile(L, "read", fnameindex, readstatus);
+        }
+        lua.lua_remove(L, fnameindex);
+        return status;
+    };
+
```
Comment (daurnimator) [inline src/lauxlib.js]: We can move these functions (`luaL_loadfile`, etc) out of the `if`

File: src/loadlib.js [hunk 5]
```diff
@@ -34,7 +36,7 @@ const AUXMARK       = [1];
 ** Returns the library; in case of error, returns NULL plus an
 ** error string in the stack.
 */
-const lsys_load = function(L, path) {
+let lsys_load = function(L, path) {
```
Comment (daurnimator) [inline src/loadlib.js]: move this inside an `else` below?

File: src/loadlib.js [hunk 6]
```diff
@@ -75,7 +93,7 @@ const noenv = function(L) {
     return b;
 };
 
-let readable = function(filename) {
```
Comment (daurnimator) [inline src/loadlib.js]: no need for function here (just declare the var)

File: src/loadlib.js [hunk 7]
```diff
@@ -95,6 +113,17 @@ if (!WEB) {
 
         return true;
     };
+} else {
+    /* TODO: use async/await ? */
+    readable = function(filename) {
+        let xhr = new XMLHttpRequest();
+        /* Following GET request done by searcher_Web will be cached */
+        xhr.open("GET", lua.to_jsstring(filename), false);
```
Comment (daurnimator) [inline src/loadlib.js]: Can be a HEAD instead of a GET.

File: src/loadlib.js [hunk 8]
```diff
@@ -430,8 +459,13 @@ const luaopen_package = function(L) {
     lauxlib.luaL_newlib(L, pk_funcs);  /* create 'package' table */
     createsearcherstable(L);
     /* set paths */
-    setpath(L, lua.to_luastring("path", true), LUA_PATH_VAR, lua.LUA_PATH_DEFAULT);
-    setpath(L, lua.to_luastring("cpath", true), LUA_CPATH_VAR, lua.LUA_CPATH_DEFAULT);
+    if (WEB) {
```
Comment (daurnimator) [inline src/loadlib.js]: no need for this branch. (see my comments above)

File: src/lua.js [hunk 9]
```diff
@@ -215,3 +215,8 @@ module.exports.lua_xmove               = lapi.lua_xmove;
 module.exports.lua_yield               = ldo.lua_yield;
 module.exports.lua_yieldk              = ldo.lua_yieldk;
 module.exports.lua_tocfunction         = lapi.lua_tocfunction;
+
+if (WEB) {
```
Comment (daurnimator) [inline src/lua.js]: no need for this (see above)

File: src/loadlib.js (hunk 7 above)
Comment (giann) [inline src/loadlib.js]: If we do the browser will effectively do 2 request (HEAD in readable, and then GET to actually get the file). If we do a GET right away, the following request will use the cache.

Comment (daurnimator) [inline src/loadlib.js]: Except in the (probably common) CORS case: the HEAD won't need a preflight, but the GET will (I think?).

File: src/lauxlib.js (hunk 3 above)
Comment (giann) [inline src/lauxlib.js]: No `lf.buff` is used by getF (node version) to read a file by chunk of 1024 bytes, but the browser version reads the whole file in one go.

Comment (daurnimator) [inline src/lauxlib.js]: Oh I didn't notice the weird indirection via `getc`

File: src/lauxlib.js [hunk 10]
```diff
@@ -644,19 +644,72 @@ const luaL_unref = function(L, t, ref) {
...
+            c: c
+        };
+    }
+};
+
+class LoadF {
+    constructor() {
+        this.n = NaN;  /* number of pre-read characters */
+        this.f = null;  /* file being read */
+        this.buff = new Buffer(1024);  /* area for reading file */
```
Comment (daurnimator) [inline src/lauxlib.js]: We should only initialise this in the `!WEB` case: otherwise this will pull in a browser shim for node's Buffer type.

File: src/lauxlib.js (hunk 3 above)
Comment (daurnimator) [inline src/lauxlib.js]: But still: various bits still use `.buff`. not only just below (`lf.buff[lf.n++] = com.c`) but also see `skipBOM`

File: src/defs.js [hunk 11]
```diff
@@ -288,14 +288,10 @@ if (WEB) {
...
-    const LUA_PATH_DEFAULT =
-        LUA_LDIR + "?.lua;" + LUA_LDIR + "?/init.lua;" +
-        LUA_CDIR + "?.lua;" + LUA_CDIR + "?/init.lua;" +
-        "./?.lua;./?/init.lua";
+    const LUA_PATH_DEFAULT = "?.lua;?/init.lua";
     module.exports.LUA_PATH_DEFAULT = LUA_PATH_DEFAULT;
 
-    const LUA_CPATH_DEFAULT =
-        LUA_CDIR + "?.js;" + LUA_CDIR + "loadall.js;./?.js";
+    const LUA_CPATH_DEFAULT = "?.js;loadall.js";
```
Comment (daurnimator) [inline src/defs.js]: Why change this?

File: src/loadlib.js (hunk 7 above)
Comment (giann) [inline src/loadlib.js]: See 71bd3b5

File: src/lauxlib.js (hunk 3 above)
Comment (giann) [inline src/lauxlib.js]: See 71bd3b5

File: src/lauxlib.js [hunk 12]
```diff
@@ -644,19 +644,144 @@ const luaL_unref = function(L, t, ref) {
...
+        }
+        let com = skipcomment(lf);
+        /* check for signature first, as we don't want to add line number corrections in binary case */
+        if (com.c === lua.LUA_SIGNATURE.charCodeAt(0) && filename) {  /* binary file? */
+            /* no need to re-open in node.js */
+        } else if (com.skipped) { /* read initial portion */
+            lf.buff[lf.n++] = '\n'.charCodeAt(0);  /* add line to correct line numbers */
+        }
+        if (com.c !== null)
+            lf.buff[lf.n++] = com.c; /* 'c' is the first character of the stream */
```
Comment (daurnimator) [inline src/lauxlib.js]: `lf.buff` does not exist at in WEB?

Comment (giann) [inline src/lauxlib.js]: yes it does see 71bd3b5

Comment (daurnimator) [inline src/lauxlib.js]: What sets/fills in lf.buff? I only see us setting `lf.f` and leaving lf.buff as an empty array.

File: src/defs.js (hunk 11 above)
Comment (giann) [inline src/defs.js]: since path are URL here, prefixing them is tricky

Comment (daurnimator) [inline src/defs.js]: howso? You could have set `LUA_CDIR` to the empty string to get the same effect as changing this line.
But even then: why have this change at all?

File: src/lauxlib.js (hunk 12 above)
Comment (giann) [inline src/lauxlib.js]: It's only used for pre-read characters (skipcomment and signature check)

**Comment by giann** [comment]
Do we want another implementation or is this mergeable ?

**Comment by daurnimator** [comment]
Could you rebase this PR so that there is less ephemeral code? (i.e. code that was added, then removed in a later commit. e.g. `LUA_WEB_PATH_DEFAULT`)

File: src/lua.js (hunk 9 above)
Comment (daurnimator) [inline src/lua.js]: this piece should be removed from this commit.

**Comment by daurnimator** [comment]
The commit 'README: Added require related section' should be removed.

File: src/defs.js [hunk 13]
```diff
@@ -293,7 +293,7 @@ if (WEB) {
         LUA_CDIR + "?.lua;" + LUA_CDIR + "?/init.lua;" +
         "./?.lua;./?/init.lua";
     module.exports.LUA_PATH_DEFAULT = LUA_PATH_DEFAULT;
-
+
```
Comment (daurnimator) [inline src/defs.js]: unrelated whitespace change?

File: src/lauxlib.js [hunk 14]
```diff
@@ -644,19 +644,144 @@ const luaL_unref = function(L, t, ref) {
...
-        }
+const errfile = function(L, what, fnameindex, error) {
+    let serr = error.message;
+    let filename = lua.lua_tostring(L, fnameindex).slice(1);
+    lua.lua_pushstring(L, lua.to_luastring(`cannot ${what} ${lua.to_jsstring(filename)}: ${serr}`));
+    lua.lua_remove(L, fnameindex);
+    return lua.LUA_ERRFILE;
+};
+
+let getc = function() {
```
Comment (daurnimator) [inline src/lauxlib.js]: Just forward declare instead of giving a dummy implementation?

File: src/lauxlib.js [hunk 15]
```diff
@@ -644,19 +644,144 @@ const luaL_unref = function(L, t, ref) {
...
+
+    getc = function(lf) {
+        return lf.pos < lf.f.length ? lf.f[lf.pos++] : null;
+    };
+
+    const luaL_loadfilex = function(L, filename, mode) {
+        let lf = new LoadF();
+        let fnameindex = lua.lua_gettop(L) + 1;  /* index of filename on the stack */
+        if (filename === null) {
+            throw new Error(`Can't read stdin in the browser`);
```
Comment (daurnimator) [inline src/lauxlib.js]: backtick strings vs double quote strings?

File: src/lauxlib.js [hunk 16]
```diff
@@ -644,19 +644,144 @@ const luaL_unref = function(L, t, ref) {
...
+        } else {
+            let jsfilename = lua.to_jsstring(filename);
+            lua.lua_pushliteral(L, `@${jsfilename}`);
+
+            let xhr = new XMLHttpRequest();
+            xhr.open("GET", jsfilename, false);
+            //xhr.responseType = "arraybuffer";
```
Comment (daurnimator) [inline src/lauxlib.js]: Do we want to keep this comment?

File: src/lauxlib.js [hunk 17]
```diff
@@ -644,19 +644,144 @@ const luaL_unref = function(L, t, ref) {
...
+            xhr.send();
+
+            if (xhr.status === 200) {
```
Comment (daurnimator) [inline src/lauxlib.js]: What about other status codes? (e.g. 203)

File: src/lauxlib.js [hunk 18]
```diff
@@ -644,19 +644,144 @@ const luaL_unref = function(L, t, ref) {
...
+        let readstatus = lf.err;
+        if (readstatus) {
+            lua.lua_settop(L, fnameindex);  /* ignore results from 'lua_load' */
+            return errfile(L, "read", fnameindex, readstatus);
+        }
+        lua.lua_remove(L, fnameindex);
+        return status;
+    };
+
+    const luaL_loadfile = function(L, filename) {
```
Comment (daurnimator) [inline src/lauxlib.js]: Can `luaL_loadfile`, `luaL_dofile`, `luaL_loadfile` be moved outside the `if(WEB)`?

File: src/lauxlib.js (hunk 16 above)
Comment (giann) [inline src/lauxlib.js]: Yes but i'll make a TODO comment to explain why

File: src/lauxlib.js [hunk 19]
```diff
@@ -771,19 +835,19 @@ if (!WEB) {
...
-    };
+const luaL_dofile = function(L, filename) {
+    return (luaL_loadfile(L, filename) || lua.lua_pcall(L, 0, lua.LUA_MULTRET, 0));
+};
 
-    module.exports.luaL_dofile    = luaL_dofile;
-    module.exports.luaL_loadfilex = luaL_loadfilex;
-    module.exports.luaL_loadfile  = luaL_loadfile;
-}
+module.exports.luaL_dofile    = luaL_dofile;
```
Comment (daurnimator) [inline src/lauxlib.js]: Move exports to end of file with the rest of them?

File: src/lauxlib.js [hunk 20]
```diff
@@ -644,19 +644,132 @@ const luaL_unref = function(L, t, ref) {
...
+    };
+
+    luaL_loadfilex = function(L, filename, mode) {
+        let lf = new LoadF();
+        let fnameindex = lua.lua_gettop(L) + 1;  /* index of filename on the stack */
+        if (filename === null) {
+            throw new Error(`Can't read stdin in the browser`);
+        } else {
+            let jsfilename = lua.to_jsstring(filename);
+            lua.lua_pushliteral(L, `@${jsfilename}`);
```
Comment (daurnimator) [inline src/lauxlib.js]: Use `lua_pushfstring` to avoid the extra js -> lua string conversion

File: README.md [hunk 21]
```diff
@@ -22,15 +22,14 @@ Lua strings are 8-bits clean and can embed `\0`. Which means that invalid UTF-8/
 To address that issue, Lua strings are represented by an array of bytes in Fengari. To push a JS string on the stack you can use `lua_pushliteral` which will convert it to an array of bytes before pushing it. To get a Lua string on the stack as a JS string you can use `lua_tojsstring` which will attempt to convert it to a UTF-16 JS string. The latter won't give you what you expect if the Lua string is not a valid UTF-16 sequence. You can also convert strings with `lua.to_luastring` and `lua.to_jsstring`.
 
 
+### `require` and `package.loadlib`
+
+In the browser `require` and `package.loadlib` try to find a file by making XHR requests.
```
Comment (daurnimator) [inline README.md]: Note that they're *synchronous* XHR requests?

File: src/lauxlib.js [hunk 22]
```diff
@@ -840,6 +901,8 @@ module.exports.luaL_gsub            = luaL_gsub;
 module.exports.luaL_len             = luaL_len;
 module.exports.luaL_loadbuffer      = luaL_loadbuffer;
 module.exports.luaL_loadbufferx     = luaL_loadbufferx;
+module.exports.luaL_loadfile  = luaL_loadfile;
```
Comment (daurnimator) [inline src/lauxlib.js]: alignment?

File: src/loadlib.js [hunk 23]
```diff
@@ -34,20 +34,37 @@ const AUXMARK       = [1];
...
-const lsys_load = function(L, path) {
-    try {
-        path = lua.to_jsstring(path);
+let lsys_load;
+if (WEB) {
+    lsys_load = function(L, path) {
+        let xhr = new XMLHttpRequest();
+        xhr.open("GET", lua.to_jsstring(path), false);
+        xhr.send();
+        /* TODO: subresource integrity check? */
```
Comment (daurnimator) [inline src/loadlib.js]: No subresource integrity check is possible here (where would the subresource hash come from?). It only makes sense when we have a `<script>`.

File: src/loadlib.js [hunk 24]
```diff
@@ -34,20 +34,37 @@ const AUXMARK       = [1];
...
+    lsys_load = function(L, path) {
+        try {
+            path = lua.to_jsstring(path);
 
-        // Relative path ?
-        if (path.startsWith('.'))
-            path = `${process.env.PWD}/${path}`;
+            // Relative path ?
+            if (path.startsWith('.'))
+                path = `${process.env.PWD}/${path}`;
```
Comment (daurnimator) [inline src/loadlib.js]: Probably doesn't belong in this PR, but this should be `process.cwd()` instead.

File: src/lauxlib.js (hunk 20 above)
Comment (giann) [inline src/lauxlib.js]: I use jsfilename with xhr.open so it's not wasted conversion

File: src/loadlib.js (hunk 24 above)
Comment (giann) [inline src/loadlib.js]: Noted

File: src/lauxlib.js (hunk 20 above)
Comment (daurnimator) [inline src/lauxlib.js]: The converstion *back* again is though.
